### PR TITLE
Added support for simulating HTML forms (including file uploads)…

### PR DESCRIPTION
**Pull request by rmcdouga**
…to the requestParameter handling. ([SLING-8424](https://issues.apache.org/jira/projects/SLING/issues/SLING-8424))

### CODE REVIEW

File: src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java [hunk 1]
```diff
@@ -516,6 +529,36 @@ public RequestParameterMap getRequestParameterMap() {
         return params;
     }
 
+    public void addRequestParameter(String name, String value) {
+    	if (this.parameterMap.containsKey(name)) {
```
Comment (stefanseifert) [inline src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java]: here (and in other lines) are tabs used instead of spaces.
please make sure to format your editor to use 4 spaces instead of tabs and reformat your changes.

File: src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java [hunk 2]
```diff
@@ -516,6 +529,36 @@ public RequestParameterMap getRequestParameterMap() {
         return params;
     }
 
+    public void addRequestParameter(String name, String value) {
```
Comment (stefanseifert) [inline src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java]: please add a bit of javadocs two this three new addRequestParameter methods

File: src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java [hunk 3]
```diff
@@ -516,6 +529,36 @@ public RequestParameterMap getRequestParameterMap() {
...
+    	if (this.parameterMap.containsKey(name)) {
+    		List<MockRequestParameter> list = Arrays.asList(this.parameterMap.get(name));
```
Comment (stefanseifert) [inline src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java]: this code is redundant across all three methods of addRequestParameter - maybe put it in one helper method?

File: src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java [hunk 4]
```diff
@@ -82,7 +83,7 @@
     private final ResourceResolver resourceResolver;
     private final RequestPathInfo requestPathInfo;
     private Map<String, Object> attributeMap = new HashMap<String, Object>();
-    private Map<String, String[]> parameterMap = new LinkedHashMap<String, String[]>();
+    private Map<String, MockRequestParameter[]> parameterMap = new LinkedHashMap<>();
```
Comment (stefanseifert) [inline src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java]: maybe we can directly use MockRequestParameterMap here?

File: src/test/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequestTest.java [hunk 5]
```diff
@@ -316,6 +317,47 @@ public void testRequestParameter() throws Exception {
         assertEquals("param1", ((MockRequestParameter)request.getRequestParameter("param1")).getName());
     }
 
+    @Test
+    public void testFormRequestParameters() throws Exception {
```
Comment (stefanseifert) [inline src/test/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequestTest.java]: maybe add a use case where a parameter with multiple values is stored using the new addRequestParameter methods.

File: src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java (hunk 1 above)
Comment (rmcdouga) [reply src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java]: I've changed the settings in my IDE (Eclipse) and reformatted the whole file.  This has caused some incidental changes (putting spaces after commas, etc.) that I think are positive so I have left them in.  The only thing that might be an issue is that my line wrapping is at 180 chars.  A couple of lines that were previously split are now merged to one line.  Is that a problem?  I assume there is a style guide somewhere.  If you could point me to it, I will set my formatting settings appropriately.

File: src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java (hunk 2 above)
Comment (rmcdouga) [reply src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java]: Done.  Update to PR is in progress.

File: src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java (hunk 3 above)
Comment (rmcdouga) [reply src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java]: Done.  This was a particularly good idea as one of the versions was correct but the other two had a bug in them.  Consolidating them made the bug obvious.  Thanks!

File: src/test/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequestTest.java (hunk 5 above)
Comment (rmcdouga) [reply src/test/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequestTest.java]: I had one such case (param4).  I could add additional cases for the other addRequestParameter() methods.  I didn't at the time time because I thought the test would be redundant.  If I had created them, then I would have spotted the issue that was highlighted by the consolidated helper function.  Now that the redundant code is consolidated however, I'm not sure if the extra tests are required (since they all just call that same helper function).

Let me know what you think.  I'm inclined not to want to do this as the extra tests are mostly just calling the same code (only one line difference between them).  That one line difference is tested by the single value tests.  Having said that, if you feel differently, I don't mind adding them in.

File: src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java (hunk 4 above)
Comment (rmcdouga) [reply src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java]: I wondered about this too.  The MockRequestParameterMap uses a HashMap while this class uses a LinkedHashMap.  I was afraid to change that since I wasn't sure if ordering of parameters is important (I haven't looked at the Sling code to see what kind of behaviour it has).  My assumption was that the original author chose a LinkedHashMap over HashMap for a reason.  Do you have any idea on why?  If I implemented this, should I make the MockRequestParameterMap use a LinkedHashMap instead?

File: src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java (hunk 1 above)
Comment (stefanseifert) [reply src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java]: i think thats fine - just make sure to make a _separate_ commit for reformatting, so it's not mixed up with other changes.

coding standards: i think sling roughly follows https://felix.apache.org/documentation/development/coding-standards.html - but that may be a bit outdated.

i think the tabs -> space conversion is the important one, not important to reformat all other parts of the file.

File: src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java (hunk 4 above)
Comment (stefanseifert) [reply src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java]: hmm, the code is using the LinkedHashMap since 2014, although i do not see a great need for it. to be on the safe side i would recommend using LinkedHashMap in the MockRequestParameterMap  as well.

File: src/test/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequestTest.java (hunk 5 above)
Comment (stefanseifert) [reply src/test/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequestTest.java]: i think it's fine if the consolidated code part is tested at least on one path.

File: src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java (hunk 1 above)
Comment (rmcdouga) [reply src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java]: >  just make sure to make a separate commit for reformatting
Sorry, I already checked in commit that has the reformatting mixed.  I think it's OK however.  The changes weren't that extensive (tabs to spaces, longer lines in 2 cases, and single spaces inserted in a few places - such as after commas and between for and open parenthesis). 

Have a look at my latest commit.  If you are concerned I can back it out and separate out the reformatting.

File: src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java (hunk 4 above)
Comment (rmcdouga) [reply src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java]: OK, will do.

Comment (rmcdouga) [reply src/main/java/org/apache/sling/servlethelpers/MockSlingHttpServletRequest.java]: So, I had a look at this and I’m not sure if this change is worth it.  Let me explain:

MockRequestParameterMap is only used in one place, getRequestParameterMap().  That method creates and returns a new copy each time it is called.

If we were to use this class for the parameter storage in MockSlingHttpRequest, then we would need to move the addRequestParameter() methods into the MockRequestParameterClass *and* retain delegate methods in the MockSlingHttpRequest class.

In my opinion, those delegate methods are the ones that people will want to call.  There aren’t any circumstances that I can think of where someone would want to create a map separately and then add parameters to it.  So the only class that will call the addRequestParameter methods will be the MockSlingRequest.  We will be adding a level of indirection that provides no benefit.

Also, if we use MockRequestParameterMap for storage, I don’t think we should return it directly from getRequestParameterMap() without making a defensive copy, so we’re not really saving anything by using it for storage.

Basically, we would be adding a bunch more code (all the delegate methods) with no real benefit.  I think I’d prefer to keep the MockRequestParameterMap strictly as a mock object returned by getRequestParameterMap() and not try and use for the internal storage mechanism too.

Let me know if you feel differently and we can discuss further.
